handler/src/main/java/io/netty/handler/ssl/ApplicationProtocolNegotiationHandler.java

```diff
@@ -0,0 +1,127 @@
+ *     private final {@link SslContext} sslCtx;
+ *
+ *     public MyInitializer({@link SslContext} sslCtx) {
+ *         this.sslCtx = sslCtx;
+ *     }
+ *
+ *     protected void initChannel({@link Channel} ch) {
+ *         {@link ChannelPipeline} p = ch.pipeline();
+ *         p.addLast(sslCtx.newHandler(...)); // Adds {@link SslHandler}
+ *         p.addLast(new MyNegotiationHandler());
+ *     }
+ * }
+ *
+ * public class MyNegotiationHandler extends {@link ApplicationProtocolNegotiationHandler} {
+ *     public MyNegotiationHandler() {
+ *         super("http/1.1");
+ *     }
+ *
+ *     protected void configurePipeline({@link ChannelHandlerContext} ctx, String protocol) {
+ *         if ("h2".equals(protocol) {
+ *             configureHttp2(ctx);
+ *         } else if ("http/1.1".equals(protocol)) {
+ *             configureHttp1(ctx);
+ *         } else {
+ *             throw new IllegalStateException("unknown protocol: " + protocol);
+ *         }
+ *     }
+ * }
+ * </pre>
+ */
+public abstract class ApplicationProtocolNegotiationHandler extends ChannelInboundHandlerAdapter {
+
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(ApplicationProtocolNegotiationHandler.class);
+
+    private final String fallbackProtocol;
+    private SslHandler sslHandler;
+    private boolean configured;
+
+    /**
+     * Creates a new instance with the specified fallback protocol name.
+     *
+     * @param fallbackProtocol the name of the protocol to use when
+     *                         ALPN/NPN negotiation fails or the client does not support ALPN/NPN
+     */
+    protected ApplicationProtocolNegotiationHandler(String fallbackProtocol) {
+        this.fallbackProtocol = ObjectUtil.checkNotNull(fallbackProtocol, "fallbackProtocol");
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        // FIXME: There is no way to tell if the SSL handler is placed before the negotiation handler.
+        final SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
+        if (sslHandler == null) {
+            throw new IllegalStateException(
+                    "cannot find a SslHandler in the pipeline (required for application-level protocol negotiation)");
+        }
+
+        this.sslHandler = sslHandler;
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (!configured && evt instanceof SslHandshakeCompletionEvent) {
+            configured = true;
+
+            SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
+            if (handshakeEvent.isSuccess()) {
+                ctx.pipeline().remove(this);
+
+                String protocol = sslHandler.applicationProtocol();
+                configurePipeline(ctx, protocol != null? protocol : fallbackProtocol);
```

Comment:
For now, the fallbackProtocol is only used for the client with no
ALPN/NPN support. Do you have a better place for this? I guess we
could move this to SslContext construction time, but I'm not 100% sure
if it's a good idea because a user might have created an SSLEngine
without using it.